src/security-testing/formal-verification.md

```diff
@@ -0,0 +1,174 @@
+---
+tags:
+  - Engineer/Developer
+  - Security Specialist
+  - Operations & Strategy
+  - Devops
+  - SRE
+contributors:
+  - role: wrote
+    users: [patrickalphac]
+---
+
+# Formal Verification
+
+Formal verification is the act of proving or disproving a given property of a system using a mathematical model. While fuzz testing tries to break properties by throwing random data at your system, formal verification tries to break properties using mathematical proofs.
+
+This is the highest level of assurance you can achieve in smart contract testing - mathematical certainty that your code behaves correctly under all possible conditions. The downside, is that it is often the most time-consuming to get correct.
+
+## What is Formal Verification?
+
+Think of formal verification as converting your code into mathematical expressions and then using powerful solvers to prove whether certain properties can be broken. Instead of testing with random inputs like fuzzing, formal verification explores **every possible execution path** mathematically.
+
+For example, if you want to prove that a function should never revert, formal verification will create mathematical representations of all possible code paths and determine if any input could cause a revert.
+
+## Symbolic Execution
+
+The most popular technique for formal verification in smart contracts is **symbolic execution**. Here's how it works:
```

Comment:
Maybe we can add `In a nutshell, symbolic execution explores **every possible execution path** mathematically.` (if we change some of the text based on the above comments)

Reply:
Great!